feat: Fast storage optimization for queries and iterations

p0mvn opened the pull request:
**Background**

Link to the [original spec](https://hackmd.io/cMctM2NsRlCIojnW3-mo5Q)

Link to the original PR [in Osmosis](https://github.com/osmosis-labs/iavl/pull/5)

**Background**
Historically IAVL has had a very slow performance during state machine execution, and for responding to queries to live state. This release speeds up these routines by an order of magnitude, alleviating large amounts of pressure from all users of the IAVL database.

**Details**

This PR introduces an auxiliary fast storage system to IAVL, which represents a copy of the latest state much more amenable to efficient querying and iteration.

Prior to this PR, all data gets & iterations suffered two significant performance drawdowns:

- Every get/iteration is forced to walk the tree structure
- Every node (including leaves) is stored on disk, with its key being a SHA256 hash of something. 

All nodes were indexed by their Merkle tree inner node hash. This breaks data locality and makes every `Get()` that should be in RAM / CPU caches instead be a random leveldb file open.

The fast storage nodes are instead indexed by the logical key on the disk. This allows us to preserve data locality for the latest state, significantly improving iterations and queries. (Depending on the particular benchmark, between 5-30x improvements) This implementation introduces a negligible overhead for writes.

**Downgrade-re-upgrade protection**
We introduced a downgrade and re-upgrade protection where we guard for potential downgrades of iavl and the subsequent enablement of the fast storage again. This is done so by storing the metadata about the current version of the storage and the latest live state stored.

**Summary of Changes**
IAVL is divided into two trees, `mutable_tree` and `immutable_tree`. Sets only happen on the mutable tree.

Things that need to change and be investigated for getting and setting, and the fast node:
- mutable tree
  - `GetVersioned`
    - [x] Change the logic to check the FastNode cache first, then do the GetImmutable logic
  - `Set`
    - [x] Cache unsaved fast nodes in memory, avoid persisting them to disk right away
    - [x] Update unsaved removals if needed
  - `Remove`
    - [x] Cache unsaved removals in memory, avoid persisting changes to disk
    - [x] Update unsaved additions if needed 
  - `SaveVersion`
    - [x] Persist unsaved fast node additions and removals to disk
    - [x] Sort by key before saving to db to ensure data locality
    - [x] Potential optimizations:
       - [x] Look into if removing before writing to db is faster
       - [x] Look into if combining sorted removals with sorted additions is faster than doing them one after the other
       - [x] Conclusion is documented here: #13  
  - `Iterate`
    - [x]  Now that we have some unsaved changes cached in the mutable tree, we cannot use the Iterate's implementation from the immutable tree. Introduce this new method.
    - [x] Ensure that we iterate in the most efficient manner in sorted order by having 2 pointers:
       1. to the next element on disk
       2. to the next unsaved change in the mutable tree
    - [x] Compare the values of the 2 pointers on each iteration and choose the next one
  - `Iterator`
     - [x] Immutable tree is embedded into mutable. From the way composition works in Go, the mutable tree can access all methods and fields of the immutable. So we need to overwrite the implementation of the immutable tree and return an invalid iterator by default.
     - [x] The iterator in the mutable tree is invalid because we cannot support updates and delayed iterations at the same time.
  - `Get`
     - [x] For the same reason as in Iterate, we must check unsaved additions first before attempting to use the strategy employed by the immutable tree
   - `enableFastStorageAndCommit` and its variations
     - [x] These are the methods that are used to perform automatic upgrades to fast storage if the system detects that fast cache is not used. This detection happens by checking the value of a special metadata key called `mstorage_version` where `m` is a new prefix. If the version is lower than the `fastStorageVersionValue` threshold - migration is triggered.
     - [x] `LoadVersion`, `LazyLoadVersion`
         * added upgrade logic and downgrade + re-upgrade protection. It works by storing the storage version in the database. 
         * if storage version is low -> upgrade
         * if the latest fast node version on disk does not match latest version in ndb -> upgrade
     - [x] isUpgradeable - determines if the upgrade sis going to be performed. This method can be called on the SDK side to determine if we should log a warning message

- immutable_tree
    - `Get` and `(GetWithIndex`
      - [x] renamed `Get` to `GetWithIndex`. `GetWithIndex` always uses the default live state traversal strategy
      - [x] introduced `Get` method. Get attempts to use the fast cache first. Only fallbacks to regular tree traversal strategy if the fast cache is disabled or tree is not of the latest version
    - `Iterator`
      - returns either regular or updated fast iterator (see more below) depending on if fast storage is enabled and migration is complete

- nodedb
    - [x] updated and tested the underlying storage management logic

- fast_iterator
    - [x] introduced and tested a new iterator that binds directly to the database iterator by searching for keys that begin with the prefix `f` which stands for fast. Basically, all fast nodes are sorted on disk by key in ascending order so we can simply traverse the disk ensuring efficient hardware access.

- unsaved_fast_iterator
    - [x] iterates over the latest state via fast nodes, taking into account unsaved changes to fast nodes that could have been made within a mutable tree

- testing
  - [x] unit tests for mutable tree Get, Set, Save, Delete
  - [x] unit tests for immutable - Get, Iterator - fast and slow
  - [x] some unit tests for  nodedb
  - [x] updated randomized tests that function like integration tests
  - [x] updated bench tests

**Old Benchmark**
Date: `2022-01-22 12:33 AM PST`
Branch: `dev/iavl_data_locality` with some modifications to the bench tests

```
go version go1.17.6 linux/amd64

Init Tree took 78.75 MB
goos: linux
goarch: amd64
pkg: github.com/cosmos/iavl/benchmarks
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkMedium/memdb-100000-100-16-40/query-no-in-tree-guarantee-slow-8         	   95841	     12315 ns/op	     592 B/op	      12 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/query-hits-slow-8                         	   85990	     15533 ns/op	     760 B/op	      15 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/iteration-slow-8                          	       2	 838458850 ns/op	88841084 B/op	 1600092 allocs/op
--- BENCH: BenchmarkMedium/memdb-100000-100-16-40/iteration-slow-8
    bench_test.go:109: completed 100000 iterations
    bench_test.go:109: completed 100000 iterations
    bench_test.go:109: completed 100000 iterations
BenchmarkMedium/memdb-100000-100-16-40/update-8                                  	   10000	    148915 ns/op	   27312 B/op	     335 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/block-8                                   	      76	  16865728 ns/op	 2910568 B/op	   35668 allocs/op
Init Tree took 46.71 MB
BenchmarkMedium/goleveldb-100000-100-16-40/query-no-in-tree-guarantee-slow-8     	   55309	     22354 ns/op	    1550 B/op	      30 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/query-hits-slow-8                     	   43566	     27137 ns/op	    2093 B/op	      39 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/iteration-slow-8                      	       1	2285116100 ns/op	225813440 B/op	 3857215 allocs/op
--- BENCH: BenchmarkMedium/goleveldb-100000-100-16-40/iteration-slow-8
    bench_test.go:109: completed 100000 iterations
BenchmarkMedium/goleveldb-100000-100-16-40/update-8                              	    6194	    307266 ns/op	   40138 B/op	     406 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/block-8                               	      28	  40663600 ns/op	 5150422 B/op	   53771 allocs/op
PASS
ok  	github.com/cosmos/iavl/benchmarks	25.797s
```

**Latest Benchmark**
Date: `2022-01-22 10:15 AM PST`
Branch: `roman/fast-node-get-set`

```
go version go1.17.6 linux/amd64

Init Tree took 114.29 MB
goos: linux
goarch: amd64
pkg: github.com/cosmos/iavl/benchmarks
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkMedium/memdb-100000-100-16-40/query-no-in-tree-guarantee-fast-8         	  672999	      1887 ns/op	     112 B/op	       4 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/query-no-in-tree-guarantee-slow-8         	   95888	     11884 ns/op	     440 B/op	       8 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/query-hits-fast-8                         	  891831	      1208 ns/op	      16 B/op	       0 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/query-hits-slow-8                         	   79842	     15644 ns/op	     607 B/op	      11 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/iteration-fast-8                          	      20	  63956090 ns/op	18400254 B/op	  300000 allocs/op
--- BENCH: BenchmarkMedium/memdb-100000-100-16-40/iteration-fast-8
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
	... [output truncated]
BenchmarkMedium/memdb-100000-100-16-40/iteration-slow-8                          	       2	 947611750 ns/op	88841044 B/op	 1600092 allocs/op
--- BENCH: BenchmarkMedium/memdb-100000-100-16-40/iteration-slow-8
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
BenchmarkMedium/memdb-100000-100-16-40/update-8                                  	    9198	    174306 ns/op	   27524 B/op	     342 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/block-8                                   	      58	  19855266 ns/op	 2948779 B/op	   36495 allocs/op
Init Tree took 66.82 MB
BenchmarkMedium/goleveldb-100000-100-16-40/query-no-in-tree-guarantee-fast-8     	  228343	      4938 ns/op	     814 B/op	      16 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/query-no-in-tree-guarantee-slow-8     	   59304	     18046 ns/op	    1420 B/op	      24 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/query-hits-fast-8                     	  611349	      1684 ns/op	      93 B/op	       1 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/query-hits-slow-8                     	   50778	     23126 ns/op	    2005 B/op	      34 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/iteration-fast-8                      	      12	  94702442 ns/op	29327220 B/op	  522988 allocs/op
--- BENCH: BenchmarkMedium/goleveldb-100000-100-16-40/iteration-fast-8
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
    bench_test.go:117: completed 100000 iterations
	... [output truncated]
BenchmarkMedium/goleveldb-100000-100-16-40/iteration-slow-8                      	       1	1716585400 ns/op	235504072 B/op	 3998006 allocs/op
--- BENCH: BenchmarkMedium/goleveldb-100000-100-16-40/iteration-slow-8
    bench_test.go:117: completed 100000 iterations
BenchmarkMedium/goleveldb-100000-100-16-40/update-8                              	    8994	    257683 ns/op	   44702 B/op	     447 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/block-8                               	      31	  44907345 ns/op	 6973362 B/op	   72924 allocs/op
PASS
ok  	github.com/cosmos/iavl/benchmarks	43.513s
```

**Benchmarks Interpretation**
Highlighting the difference in performance from the latest benchmarks:

Old branch is `dev/iavl_data_locality`
New branch is `roman/fast-node-get-set`

Initial size: `100,000` key-val pairs
Block size: `100` keys
Key length: `16` bytes
Value length: `40` bytes

**Query with no guarantee of the key being in the tree**:
- Old: `22354 ns/op`
- New on regular logic: `18046 ns/op`
- New on fast logic: `4938 ns/op`
- New fast logic  shows a **77% decrease** in time relative to the old branch 

**Query with the key guaranteed to be in the latest tree**:
- Old: ` 27137 ns/op`
- New on regular logic: `23126 ns/op`
- New on fast logic: `1684 ns/op`
- New fast logic  shows a **93% decrease** in time relative to the old branch 

**Iteration**:
- Old: `2285116100 ns/op`
- New on old logic: `1716585400 ns/op`
- New on fast logic: `94702442 ns/op`
- New fast logic shows a **96% decrease** in time relative to the old branch 

**Update**:
run Set, if this is a try that is divisible by blockSize, attempt to SaveVersion and if the latest saved version number history exceeds 20, delete the oldest version
- Old: `307266 ns/op`
- New: `257683 ns/op`
- New logic shows a **16% decrease** in time relative to the old branch

**Block**: 
for block size, run Get and Set. At the end of the block, SaveVersion and if the latest saved version number history exceeds 20, delete the oldest version
- Old: `40663600 ns/op`
- New: `44907345 ns/op`
- New logic shows a **9% increase** in time relative to the old branch

odeke-em commented on unsaved_fast_iterator.go:
```diff
@@ -0,0 +1,233 @@
+	iter := &UnsavedFastIterator{
+		start:                    start,
+		end:                      end,
+		err:                      nil,
```
Unnecessary, the zero value for an error is nil.

odeke-em reviewed (commented):
Thank you for this change @p0mvn and thank you for the tag @marbar3778! @p0mvn I've added my first round of review, please take a look.

p0mvn commented:
@odeke-em Thanks for the feedback. I will address all comments and fix the benchmarks soon

p0mvn commented on mutable_tree.go:
```diff
@@ -439,6 +506,74 @@ func (tree *MutableTree) LoadVersionForOverwriting(targetVersion int64) (int64,
+		// downgrade and subsequent re-upgrade, we cannot know for sure which fast nodes have been removed while downgraded,
+		// Therefore, there might exist stale fast nodes on disk. As a result, to avoid persisting the stale state, it might
+		// be worth to delete the fast nodes from disk.
+		fastItr := NewFastIterator(nil, nil, true, tree.ndb)
```
What is the purpose of duplicating the `Close()` below if we are doing it with defer?

p0mvn replied:
Nvm, I understand

p0mvn commented on repair_test.go:
```diff
@@ -14,6 +14,8 @@ import (
 )
 
 func TestRepair013Orphans(t *testing.T) {
+	t.Skip()
```
That was accidental, thanks for catching that

p0mvn commented on testutils_test.go:
```diff
@@ -115,6 +123,187 @@ func expectTraverse(t *testing.T, trav traverser, start, end string, count int)
+}
+
+func getRandomKeyFrom(mirror map[string]string) string {
+	keys := make([]string, 0, len(mirror))
+	for k := range mirror {
+		keys = append(keys, k)
+	}
+	key := keys[rand.Intn(len(keys))]
+	return key
```
Great suggestion, I took the go map iteration approach

p0mvn commented on testutils_test.go:
```diff
@@ -115,6 +123,187 @@ func expectTraverse(t *testing.T, trav traverser, start, end string, count int)
+}
+
+func setupMirrorForIterator(t *testing.T, config *iteratorTestConfig, tree *MutableTree) [][]string {
+	mirror := make([][]string, 0)
```
Went for the first one since performance here does not matter. However, other people could get confused in the future as for why we assumed a certain number

p0mvn commented on tree_random_test.go:
```diff
@@ -389,9 +401,53 @@ func assertMirror(t *testing.T, tree *MutableTree, mirror map[string]string, ver
+	for key, cacheElem := range tree.ndb.fastNodeCache {
+		liveFastNode := mirror[key]
+
+		require.NotNil(t, liveFastNode, "cached fast node must be in live tree")
```
```
liveFastNode, ok := mirror[key]

require.True(t, ok, "cached fast node must be in the live tree")
```

p0mvn commented:
@odeke-em Thanks so much for such a great and detailed review.

I addressed all your comments. The current TODOs left for me in this PR are the following:
- [ ] resolve all linter issues
- [ ] better benchstat summary
- [ ] Look into the failing bench test - related to `cleveldb` and might be due to a recent tendermint upgrade

Please note that some of the benchmarks that are worse-off are expected. Optimizing for reads forces us to introduce some additional overhead for writes. I will get back with a more detailed report on every negative change:
- [ ] document all performance regressions

alexanderbez reviewed (approved):
yoloACK

p0mvn commented:
Better benchstat summary:

```
~/projects/iavl (master)$ benchstat bench_old_large.log bench_fast_large.log
name                                                             old time/op    new time/op    delta
Large/memdb-1000000-100-16-40/query-no-in-tree-guarantee-16        11.6µs ±18%     2.1µs ± 6%  -81.86%  (p=0.008 n=5+5)
Large/memdb-1000000-100-16-40/query-hits-16                        13.0µs ± 3%     1.5µs ±21%  -88.06%  (p=0.008 n=5+5)
Large/memdb-1000000-100-16-40/iteration-16                          7.05s ± 9%     0.52s ± 2%  -92.67%  (p=0.008 n=5+5)
Large/memdb-1000000-100-16-40/update-16                             220µs ±10%     218µs ± 9%     ~     (p=0.841 n=5+5)
Large/memdb-1000000-100-16-40/block-16                             23.0ms ± 3%    24.3ms ± 2%   +5.28%  (p=0.016 n=5+5)
Large/goleveldb-1000000-100-16-40/query-no-in-tree-guarantee-16    21.5µs ±11%     5.7µs ± 2%  -73.60%  (p=0.008 n=5+5)
Large/goleveldb-1000000-100-16-40/query-hits-16                    21.8µs ± 5%     2.7µs ±28%  -87.75%  (p=0.008 n=5+5)
Large/goleveldb-1000000-100-16-40/iteration-16                      16.6s ±11%      0.5s ± 1%  -96.81%  (p=0.008 n=5+5)
Large/goleveldb-1000000-100-16-40/update-16                         245µs ±11%     269µs ±10%     ~     (p=0.222 n=5+5)
Large/goleveldb-1000000-100-16-40/block-16                         28.8ms ± 9%    31.8ms ±19%     ~     (p=0.151 n=5+5)

name                                                             old alloc/op   new alloc/op   delta
Large/memdb-1000000-100-16-40/query-no-in-tree-guarantee-16          566B ±37%       88B ± 0%  -84.45%  (p=0.008 n=5+5)
Large/memdb-1000000-100-16-40/query-hits-16                          606B ± 0%       67B ±46%  -88.91%  (p=0.008 n=5+5)
Large/memdb-1000000-100-16-40/iteration-16                          888MB ± 0%     184MB ± 0%  -79.28%  (p=0.016 n=4+5)
Large/memdb-1000000-100-16-40/update-16                            36.7kB ± 3%    37.2kB ± 3%     ~     (p=0.222 n=5+5)
Large/memdb-1000000-100-16-40/block-16                             3.80MB ± 1%    3.86MB ± 1%   +1.62%  (p=0.008 n=5+5)
Large/goleveldb-1000000-100-16-40/query-no-in-tree-guarantee-16    2.72kB ±29%    0.99kB ± 0%  -63.51%  (p=0.008 n=5+5)
Large/goleveldb-1000000-100-16-40/query-hits-16                    2.70kB ± 1%    0.34kB ±46%  -87.34%  (p=0.008 n=5+5)
Large/goleveldb-1000000-100-16-40/iteration-16                     3.71GB ± 0%    0.30GB ± 0%  -92.00%  (p=0.016 n=4+5)
Large/goleveldb-1000000-100-16-40/update-16                        79.8kB ± 6%    84.1kB ±10%     ~     (p=0.421 n=5+5)
Large/goleveldb-1000000-100-16-40/block-16                         9.05MB ± 9%    9.82MB ±16%     ~     (p=0.095 n=5+5)

name                                                             old allocs/op  new allocs/op  delta
Large/memdb-1000000-100-16-40/query-no-in-tree-guarantee-16          10.8 ±39%       3.0 ± 0%  -72.22%  (p=0.008 n=5+5)
Large/memdb-1000000-100-16-40/query-hits-16                          11.0 ± 0%       1.0 ± 0%  -90.91%  (p=0.000 n=5+4)
Large/memdb-1000000-100-16-40/iteration-16                          16.0M ± 0%      3.0M ± 0%  -81.25%  (p=0.016 n=4+5)
Large/memdb-1000000-100-16-40/update-16                               466 ± 8%       459 ± 8%     ~     (p=1.000 n=5+5)
Large/memdb-1000000-100-16-40/block-16                              52.0k ± 0%     53.2k ± 0%   +2.29%  (p=0.008 n=5+5)
Large/goleveldb-1000000-100-16-40/query-no-in-tree-guarantee-16      37.2 ±29%      19.0 ± 0%  -48.92%  (p=0.008 n=5+5)
Large/goleveldb-1000000-100-16-40/query-hits-16                      36.0 ± 0%       6.0 ±50%  -83.33%  (p=0.008 n=5+5)
Large/goleveldb-1000000-100-16-40/iteration-16                      49.8M ± 0%      5.3M ± 0%  -89.34%  (p=0.016 n=4+5)
Large/goleveldb-1000000-100-16-40/update-16                           684 ± 6%       729 ±12%     ~     (p=0.310 n=5+5)
Large/goleveldb-1000000-100-16-40/block-16                          81.5k ± 7%     84.8k ±12%     ~     (p=0.421 n=5+5)
```

Since the API has changes and bench tests were rewritten between `master` and the current branch, I had to create a custom version of `bench_test.go` that works for both branches.

p0mvn commented:
The bencher was failing from an earlier commit and is now broken on master.

From initial investigation, it only happens for databases other than `goleveldb` and `memdb` so I commented them out.

I'm going to let bencher pass for `goleveldb` and `memdb`.

For all others, #469 is created. To be addressed in another PR.

p0mvn commented:
An update from me: Osmosis validators reported an increased RAM usage while upgrading to fast storage. I've been digging into that. So far, I don't see any issues besides just naturally requiring more RAM to create a copy of the live state. I have some optimization ideas and will document them as soon as the investigation is done.

I've been mostly doing heap profiles. Now, I want to make sure that there is no memory leak due to a blocked thread somewhere. However, this isn't likely since validators only reported the increased RAM during the upgrade. Eventually, it goes down. Again, the increased RAM usage during the upgrade is expected though it can be optimized if that's too bad of an issue.

I heard reports of getting OOM killed with 16GB RAM node in Osmosis, still need to try testing that myself.

TODOs left in this PR:
- [ ] look into gobencher timing out with large data
- [ ] document all performance regressions
  * everything I observed so far is expected, just need some spare cycles to document it for the reviewers

robert-zaremba commented:
Is the FastNode DB only in the RAM, or you store it in DM? I think we need the latter, and clean the cache in RAM every x blocks. Otherwise it will dramatically change the validator requirements.

p0mvn commented:
No, it is primarily in the DM. There is some configurable cache in `nodedb`. Also, some additions and removals are persisted in memory until we `SaveVersion` them to DM.

We have observed some RAM increase. However, nothing notable so far.

There were issues with RAM increasing during the upgrade of the storage to the point where it would get OOM'ed on 16 Gb nodes.
However, I have a spike to address that here: https://github.com/osmosis-labs/iavl/pull/19

It does the following:
- force GC every second
- avoid using the `nodedb` cache during the upgrade

p0mvn commented:
Added several data race fixes that have been identified by running nodes in Osmosis.

This PR is updated with the latest changes and caught up with the `master` branch.

My local tests and linter are passing. Could someone run the CI please? I don't have the permissions.

Also, gobencher is failing for reasons outside of my control: https://github.com/cosmos/iavl/pull/468/checks?check_run_id=5726930260

marbar3778 commented:
thanks for the update, sorry for the delay I was off last week. We will get to this asap

p0mvn commented:
> Better benchstat summary:
> 
> ```
> ~/projects/iavl (master)$ benchstat bench_old_large.log bench_fast_large.log
> name                                                             old time/op    new time/op    delta
> Large/memdb-1000000-100-16-40/query-no-in-tree-guarantee-16        11.6µs ±18%     2.1µs ± 6%  -81.86%  (p=0.008 n=5+5)
> Large/memdb-1000000-100-16-40/query-hits-16                        13.0µs ± 3%     1.5µs ±21%  -88.06%  (p=0.008 n=5+5)
> Large/memdb-1000000-100-16-40/iteration-16                          7.05s ± 9%     0.52s ± 2%  -92.67%  (p=0.008 n=5+5)
> Large/memdb-1000000-100-16-40/update-16                             220µs ±10%     218µs ± 9%     ~     (p=0.841 n=5+5)
> Large/memdb-1000000-100-16-40/block-16                             23.0ms ± 3%    24.3ms ± 2%   +5.28%  (p=0.016 n=5+5)
> Large/goleveldb-1000000-100-16-40/query-no-in-tree-guarantee-16    21.5µs ±11%     5.7µs ± 2%  -73.60%  (p=0.008 n=5+5)
> Large/goleveldb-1000000-100-16-40/query-hits-16                    21.8µs ± 5%     2.7µs ±28%  -87.75%  (p=0.008 n=5+5)
> Large/goleveldb-1000000-100-16-40/iteration-16                      16.6s ±11%      0.5s ± 1%  -96.81%  (p=0.008 n=5+5)
> Large/goleveldb-1000000-100-16-40/update-16                         245µs ±11%     269µs ±10%     ~     (p=0.222 n=5+5)
> Large/goleveldb-1000000-100-16-40/block-16                         28.8ms ± 9%    31.8ms ±19%     ~     (p=0.151 n=5+5)
> 
> name                                                             old alloc/op   new alloc/op   delta
> Large/memdb-1000000-100-16-40/query-no-in-tree-guarantee-16          566B ±37%       88B ± 0%  -84.45%  (p=0.008 n=5+5)
> Large/memdb-1000000-100-16-40/query-hits-16                          606B ± 0%       67B ±46%  -88.91%  (p=0.008 n=5+5)
> Large/memdb-1000000-100-16-40/iteration-16                          888MB ± 0%     184MB ± 0%  -79.28%  (p=0.016 n=4+5)
> Large/memdb-1000000-100-16-40/update-16                            36.7kB ± 3%    37.2kB ± 3%     ~     (p=0.222 n=5+5)
> Large/memdb-1000000-100-16-40/block-16                             3.80MB ± 1%    3.86MB ± 1%   +1.62%  (p=0.008 n=5+5)
> Large/goleveldb-1000000-100-16-40/query-no-in-tree-guarantee-16    2.72kB ±29%    0.99kB ± 0%  -63.51%  (p=0.008 n=5+5)
> Large/goleveldb-1000000-100-16-40/query-hits-16                    2.70kB ± 1%    0.34kB ±46%  -87.34%  (p=0.008 n=5+5)
> Large/goleveldb-1000000-100-16-40/iteration-16                     3.71GB ± 0%    0.30GB ± 0%  -92.00%  (p=0.016 n=4+5)
> Large/goleveldb-1000000-100-16-40/update-16                        79.8kB ± 6%    84.1kB ±10%     ~     (p=0.421 n=5+5)
> Large/goleveldb-1000000-100-16-40/block-16                         9.05MB ± 9%    9.82MB ±16%     ~     (p=0.095 n=5+5)
> 
> name                                                             old allocs/op  new allocs/op  delta
> Large/memdb-1000000-100-16-40/query-no-in-tree-guarantee-16          10.8 ±39%       3.0 ± 0%  -72.22%  (p=0.008 n=5+5)
> Large/memdb-1000000-100-16-40/query-hits-16                          11.0 ± 0%       1.0 ± 0%  -90.91%  (p=0.000 n=5+4)
> Large/memdb-1000000-100-16-40/iteration-16                          16.0M ± 0%      3.0M ± 0%  -81.25%  (p=0.016 n=4+5)
> Large/memdb-1000000-100-16-40/update-16                               466 ± 8%       459 ± 8%     ~     (p=1.000 n=5+5)
> Large/memdb-1000000-100-16-40/block-16                              52.0k ± 0%     53.2k ± 0%   +2.29%  (p=0.008 n=5+5)
> Large/goleveldb-1000000-100-16-40/query-no-in-tree-guarantee-16      37.2 ±29%      19.0 ± 0%  -48.92%  (p=0.008 n=5+5)
> Large/goleveldb-1000000-100-16-40/query-hits-16                      36.0 ± 0%       6.0 ±50%  -83.33%  (p=0.008 n=5+5)
> Large/goleveldb-1000000-100-16-40/iteration-16                      49.8M ± 0%      5.3M ± 0%  -89.34%  (p=0.016 n=4+5)
> Large/goleveldb-1000000-100-16-40/update-16                           684 ± 6%       729 ±12%     ~     (p=0.310 n=5+5)
> Large/goleveldb-1000000-100-16-40/block-16                          81.5k ± 7%     84.8k ±12%     ~     (p=0.421 n=5+5)
> ```
> 
> Since the API has changes and bench tests were rewritten between `master` and the current branch, I had to create a custom version of `bench_test.go` that works for both branches.

@robert-zaremba I will do a new one in the next few days, sorry for the delay. However, here's the latest one.

lyh169 commented:
![screenshot-20220725-180014](https://user-images.githubusercontent.com/44960676/180751569-add5d07c-964f-43ab-b69a-c24c19591cbb.png)

hi, why modify  mtx from sync.RWMutex to sync.Mutex type . I think the sync.RWMutex should also work

robert-zaremba commented:
@lyh169 there was a benchmark suggesting that. (normal Mutex was faster)

giskook commented on mutable_tree.go:
```diff
@@ -26,14 +28,16 @@ var ErrVersionDoesNotExist = errors.New("version does not exist")
+	orphans                  map[string]int64       // Nodes removed by changes to working tree.
+	versions                 map[int64]bool         // The previous, saved versions of the tree.
+	allRootLoaded            bool                   // Whether all roots are loaded or not(by LazyLoadVersion)
+	unsavedFastNodeAdditions map[string]*FastNode   // FastNodes that have not yet been saved to disk
```
Why the unsavedFastNodeAdditions and unsavedFastNodeRemovals do not need mtx to protect?  the map would not be concurrent access?

giskook commented on mutable_tree.go:
```diff
@@ -26,14 +28,16 @@ var ErrVersionDoesNotExist = errors.New("version does not exist")
+	versions                 map[int64]bool         // The previous, saved versions of the tree.
+	allRootLoaded            bool                   // Whether all roots are loaded or not(by LazyLoadVersion)
+	unsavedFastNodeAdditions map[string]*FastNode   // FastNodes that have not yet been saved to disk
+	unsavedFastNodeRemovals  map[string]interface{} // FastNodes that have not yet been removed from disk
```
Hi, I review the code that the `unsavedFastNodeRemovals`' value is only bool. so why not use `map[string]struct{}`? @p0mvn

robert-zaremba replied:
that is a good optimization

robert-zaremba replied:
that is a good optimization

robert-zaremba commented:
wow, I've just realized that I had not submitted comments.
